### DISCUSSION TITLE
Question on how the tool works

### DISCUSSION THREAD

**Discussion by bogdan-ionut** [xob0t/Google-Photos-Toolkit#24] [Q&A] [answered]
Hi there! I can't emphasize enough how useful and awesome this tool is. It is just AMAZING!!! Thanks a lot for taking the time to develop such a wonderful tool.

I am a software developer myself and I tried to use the Google Photos API to do smth like this but I could not find information regarding the media items storage in the metadata.

I am left wondering now how were you able to find the photos that take up storage? Just the bigger picture on how does the tools identifies that.

Thanks a lot!

**Comment by xob0t** [comment, marked as answer]
Hi!
This tool uses the same api used by https://photos.google.com/
The api is undocumented and uses a data format akin to protobuf.
So i've spent some time reverse-engineering it, and this is the result.
The process is to trigger an action in the UI, find the request that was sent, and try to figure out the response. Response data is readable, but it is just values with names stripped, so some guesswork is involved. I used burp's response mocking in a couple of tough places.

Here is the code for sending api requests https://github.com/xob0t/Google-Photos-Toolkit/blob/main/src/api/api.js
And here you can find functions for parsing the api responses https://github.com/xob0t/Google-Photos-Toolkit/blob/main/src/api/parser.js

Also, i've made another userscript that logs all api responses to console in a more readable way https://github.com/xob0t/Google-Photos-Toolkit/blob/main/tools/response_parser.user.js

**Reply by xob0t** [reply]
Oh, and thanks for the kind words, i'm glad you like it!
